nflog/nflog.go

```diff
@@ -0,0 +1,566 @@
+	st  gossipData
+}
+
+// Option configures a new Log implementation.
+type Option func(*nlog) error
+
+// WithMesh registers the log with a mesh network with which
+// the log state will be shared.
+func WithMesh(create func(g mesh.Gossiper) mesh.Gossip) Option {
+	return func(l *nlog) error {
+		l.gossip = create(l)
+		return nil
+	}
+}
+
+// WithRetention sets the retention time for log st.
+func WithRetention(d time.Duration) Option {
+	return func(l *nlog) error {
+		l.retention = d
+		return nil
+	}
+}
+
+// WithNow overwrites the function used to retrieve a timestamp
+// for the current point in time.
+// This is generally useful for injection during tests.
+func WithNow(f func() time.Time) Option {
+	return func(l *nlog) error {
+		l.now = f
+		return nil
+	}
+}
+
+func WithLogger(logger log.Logger) Option {
+	return func(l *nlog) error {
+		l.logger = logger
+		return nil
+	}
+}
+
+// WithMaintenance configures the Log to run garbage collection
+// and snapshotting to the provided file at the given interval.
+// On startup the a snapshot is also loaded from the given file.
+//
+// The maintenance terminates on receiving from the provided channel.
+// The done function is called after the final snapshot was completed.
+//
+// Providing a 0 duration will not run background processing.
+// Providing an empty file name will skip snapshotting.
+func WithMaintenance(sf string, d time.Duration, stopc chan struct{}, done func()) Option {
+	return func(l *nlog) error {
+		l.runInterval = d
+		l.stopc = stopc
+		l.done = done
+		l.snapf = sf
+		return nil
+	}
+}
+
+func utcNow() time.Time {
+	return time.Now().UTC()
+}
+
+// New creates a new notification log based on the provided options.
+// The snapshot is loaded into the Log if it is set.
+func New(opts ...Option) (Log, error) {
+	l := &nlog{
+		logger: log.NewNopLogger(),
+		now:    utcNow,
+		st:     map[string]*pb.MeshEntry{},
+	}
+	for _, o := range opts {
+		if err := o(l); err != nil {
+			return nil, err
+		}
+	}
+	if l.snapf != "" {
+		f, err := os.Open(l.snapf)
+		if err != nil {
+			return l, err
+		}
+		defer f.Close()
+
+		if err := l.loadSnapshot(f); err != nil {
+			return l, err
+		}
+	}
+	go l.run()
+
+	return l, nil
+}
+
+// run periodic background maintenance.
+func (l *nlog) run() {
+	if l.runInterval == 0 || l.stopc == nil {
+		return
+	}
+	t := time.NewTicker(l.runInterval)
+	defer t.Stop()
+
+	if l.done != nil {
+		defer l.done()
+	}
+
+	f := func() error {
+		if _, err := l.GC(); err != nil {
+			return err
+		}
+		if l.snapf == "" {
+			return nil
+		}
+		f, err := openReplace(l.snapf)
+		if err != nil {
+			return err
+		}
+		// TODO(fabxc): potentially expose snapshot size in log message.
+		if _, err := l.Snapshot(f); err != nil {
+			return err
+		}
+		return f.Close()
+	}
+
+	for {
+		select {
+		case <-l.stopc:
+			return
+		case <-t.C:
+			if err := f(); err != nil {
+				l.logger.Log("msg", "running maintenance failed", "err", err)
+			}
+		}
+	}
```

Comment:
Code below is unreachable (the only way out of the for loop above is the return in L220).